lib/srv/tcpip/forward.go

```diff
@@ -0,0 +1,218 @@
+// success/failure in dialing the target address.
+type ForwardResponse struct {
+	// Success is true if the target address was successfully dialed.
+	Success bool `json:"success"`
+	// Error is the error string associated with a failed dial.
+	Error string `json:"error"`
+}
+
+// NewDialer wraps a [utils.SocketPairDialer], sending dials as messages understood by the
+// corresponding [Forwarder].  Using this type in conjunction with
+// the socket pair dialer/listener helpers from utils makes it easy to forward arbitrary
+// TCP dials via a socket.
+func NewDialer(inner utils.SocketPairDialer) *Dialer {
+	return &Dialer{
+		inner: inner,
+	}
+}
+
+// Dialer is a tcp dialer that forwards dials across a socket.
+type Dialer struct {
+	inner utils.SocketPairDialer
+}
+
+// Dial dials the target address.
+func (d *Dialer) Dial(addr string) (net.Conn, error) {
+	conn, err := d.inner.Dial()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	req := ForwardRequest{
+		DestinationAddress: addr,
+	}
+
+	reqBytes, err := utils.FastMarshal(&req)
+	if err != nil {
+		conn.Close()
+		return nil, trace.Errorf("failed to marshal forward request: %v", err)
+	}
+
+	if err := WriteLengthPrefixedMessage(conn, reqBytes); err != nil {
+		conn.Close()
+		return nil, trace.Errorf("failed to send forward request: %v", err)
+	}
+
+	rspBytes, err := ReadLengthPrefixedMessage(conn)
+	if err != nil {
+		conn.Close()
+		return nil, trace.Errorf("failed to read forward response: %v", err)
+	}
+
+	var rsp ForwardResponse
+	if err := utils.FastUnmarshal(rspBytes, &rsp); err != nil {
+		conn.Close()
+		return nil, trace.Errorf("failed to unmarshal forward response: %v", err)
+	}
+
+	if !rsp.Success {
+		conn.Close()
+		return nil, trace.Errorf("failed to dial forwarding target: %q", err)
+	}
+
+	// connection is now ready to forward
+	return conn, nil
+}
+
+// Close closes the inner dialer.
+func (d *Dialer) Close() error {
+	d.inner.Close()
+	return nil
+}
+
+// Forwarder implements the core logic of the forwarding child process. It is designed to
+// sit behind a socket, serving tcpip forward requests passed to it from the parent.
+type Forwarder struct {
+	listener     net.Listener
+	closeContext context.Context
+	cancel       context.CancelCauseFunc
+}
+
+// NewForwarder builds a forwarder. The passed in file descriptor must point to a valid
+// listener (presumably passed in from the parent process). The inactivityTimeout determines
+// roughly how long the forwarder should continue listening for incoming connections after
+// the last time it was active.
+func NewForwarder(listener net.Listener) *Forwarder {
+	ctx, cancel := context.WithCancelCause(context.Background())
+	return &Forwarder{
+		listener:     listener,
+		closeContext: ctx,
+		cancel:       cancel,
+	}
+}
+
+// errForwarderClosed is used to indicate termination due to explicit closure.
+var errForwarderClosed = errors.New("socket forwarder closed")
```

Comment:
this error is just a sentinel value used so that some of the log lines can recognize and ignore deliberate closure, so `errors.New` feels appropriate to me.